lib/vssh/libssh.c

```diff
@@ -336,7 +336,13 @@ static int myssh_is_known(struct connectdata *conn)
   struct curl_khkey knownkey;
 #endif
 
-#if LIBSSH_VERSION_INT >= SSH_VERSION_INT(0,8,0)
+  /* libssh 0.8.0 renamed ssh_get_publickey and marked it as deprecated.
+   * Ubuntu 18.04 however ships with a pre-release version (commit 94fa1e38)
+   * that also includes that change. As workaround, check for a symbol from
+   * libssh/sftp.h that was added after 0.7.0 to avoid deprecation warnings.
+   */
+#if (LIBSSH_VERSION_INT >= SSH_VERSION_INT(0, 8, 0)) || \
+    (LIBSSH_VERSION_INT == SSH_VERSION_INT(0, 7, 0) && defined(SSH_S_IFMT))
```

Comment:
Wouldn't be better if the existence of the new symbol was checked during configuration and used if defined?

Maybe cmake's [CheckSymbolExists](https://cmake.org/cmake/help/v3.2/module/CheckSymbolExists.html) or [CheckFunctionExists](https://cmake.org/cmake/help/v3.2/module/CheckFunctionExists.html) could be used.

For autoconf it would require writing m4 macros to detect the presence of the new symbol.

Reply:
Yes, except for builds using other methods than autotools and cmake but I'm pretty sure they're not using libssh to any large extent...

Reply:
I'd prefer less feature detection checks in autotools/cmake rather than more since they slow down the configure/cmake step. Luckily this case can easily be detected by these preprocessor macros, so no symbol checks are needed.

Reply:
Hi @ansasaki, do you have any objections on this patch? It is a blocker before #5370 can be merged.

Reply:
Hi @Lekensteyn, I just think the workaround for the check is a bit "hacky", but if it works for you I won't object. Since it is checking for something that is not going to change, I believe it wouldn't create problems for the future. As you mentioned, the workaround does not increase the configuration time and avoids having to write checks for cmake, autotools, and a fall back when none of them are available, which can be considered arguments in favour of it.

Reply:
I agree it is a bit hacky, but I thought it is the best solution available to ensure that CI does not fail.

Some of the options are:

- Special-case a known bad version and detect it with macros. Unlikely to break older or newer versions since the version check is fixed. Implemented by this patch.
- Set `-Wno-error` or `-Wno-deprecation` for the affected file. This may hide other errors in the same file.
- Add even more lines of code to configure.ac/CMakeLists.txt to perform the symbol check. May be more complex than worth it.
- Ask Ubuntu developers to update their packages. Since both functions are functionally equivalent and it is "just" a compiler warning, I don't see them pushing a "fix" anytime soon.
- Accept that libssh on Ubuntu 18.04 is an uncommon configuration. libssh2 is available and so are newer Ubuntu versions, including 20.04. To fix CI, patch the header as workaround: `sed '/ ssh_get_publickey(/s/^SSH_DEPRECATED //' /usr/include/libssh/libssh.h -i`. Unfortunately, this patch would be needed for both Travis CI and Azure Pipelines as they only offer 18.04 as latest option ([Travis CI - Linux][1], [Azure ubuntu-latest][2]).

Given these, it seems that the first option achieves the desired objective (no CI failures) with the least complexity.

 [1]: https://docs.travis-ci.com/user/reference/overview/#linux
 [2]: https://docs.microsoft.com/en-us/azure/devops/pipelines/agents/hosted?view=azure-devops#use-a-microsoft-hosted-agent